### DISCUSSION TITLE
Tips for removing overlapping notes within the same channel

### DISCUSSION THREAD

**Discussion by Jenix-Park** [melanchall/drywetmidi#187] [Q&A] [answered]
This is my last question.
I have some MIDI files that contain many overlapped notes in the same channel.
I can't share them without the authors' permission, so I copied a small part. See [this file](https://drive.google.com/file/d/1fpK-IF5xYK8QRgiDWEBjXLPQA2H2nXPL/view?usp=sharing), please.
In this specific case, overlapped (the same note number) notes' durations are all zero.
So I can solve the problem by ignoring notes with a length of 0.
But there may be other complicated cases.
For some reason, many short notes close to 1 may be used here and there.
At the same time, there may be overlapped notes like the file attached here, and their lengths are similar to those short notes.
I want to give users an option to filter out overlapped short notes in the same channel for this kind of case.
I mean, I want an option to choose longer notes only when there are multiple overlapping ones at the exact same time in the same channel.

Is there any built-in way or tips I can easily achieve this?

**Comment by melanchall** [comment, marked as answer]
Hi,

Unfortunately there is no built-in tool to solve the task. You need to iterate over notes manually and check whether they overlap another notes or not. So, for example, if we have notes `A` and `B`, then `B` overlaps `A` if:

1. `B.Time` greater than or equal to `A.Time` and
2. `B.Time + B.Length` less than or equal to `A.Time + A.Length`.

**Reply by Jenix-Park** [reply]
Thank you, I'll try it myself :)
Update: Your "Notes merging" example code from [this page](https://www.codeproject.com/Articles/1200014/DryWetMIDI-High-level-processing-of-MIDI-files#top) is somewhat similar to what I want to solve.

By the way, may I ask something not related to your library but music theory?
If I'm not mistaken, you are also a musician, so I thought you are the right person to ask.
Is there any easy algorithm or something (other than using a predefined table) to automatically determine the chord name from a user's three or more notes, as some DAWs and keyboards do?

**Reply by melanchall** [reply]
[Notes merger](https://melanchall.github.io/drywetmidi/articles/tools/Notes-merger.html) won't work in general case for your task. Yes, it will merge overlapping notes into one, BUT it also may merge two notes with no pause between them. But such notes are not overlapping so I think it will create troubles.

As for getting a chord's name: I tried different algorithms, but they all were not easy. Also my algorithms were pretty slow on large collections of chords. So I've implemented simple solution in my library: [search by chords names table (CNT)](https://melanchall.github.io/drywetmidi/articles/music-theory/Chord.html#chords-names-table). Idea is to search for collection of offsets from root note. So you can use [GetNames](https://melanchall.github.io/drywetmidi/api/Melanchall.DryWetMidi.MusicTheory.Chord.html#Melanchall_DryWetMidi_MusicTheory_Chord_GetNames) method to know the name of a chord if it's supported in CNT.

**Reply by Jenix-Park** [reply]
Yeah, I know what the merger does :)
As I said earlier, I've been reading all the documents you've written to fully understand.
Because you wrote it a long time ago, so you might have forgotten about it, but you have the "Notes merging" example code on the "DryWetMIDI: High-Level Processing of MIDI Files" page.
It doesn't use any mergers and it only merges overlapping ones in the same channel.
I haven't tried it yet, but it seems like a way to go for me just by looking at it.
Anyway, the documents are very helpful in many ways.
You've done a great job!

I'm thinking of adding a feature to show the chord's name of user input in real-time. So performance is really important. Thanks for the advice!
